pkg/reconciler/image/build_needed.go

```diff
@@ -1,85 +1,156 @@
+	srcResolver *v1alpha1.SourceResolver,
+	builder v1alpha1.BuilderResource) *BuildDeterminer {
+
+	return &BuildDeterminer{
+		img:         img,
+		lastBuild:   lastBuild,
+		srcResolver: srcResolver,
+		builder:     builder,
 	}
+}
 
-	if lastBuild == nil || im.Spec.Tag != lastBuild.Tag() {
-		return []string{v1alpha1.BuildReasonConfig}, corev1.ConditionTrue
-	}
+type BuildDeterminer struct {
+	img         *v1alpha1.Image
+	lastBuild   *v1alpha1.Build
+	srcResolver *v1alpha1.SourceResolver
+	builder     v1alpha1.BuilderResource
 
-	var reasons []string
+	changeSummary buildchange.ChangeSummary
+}
 
-	if sourceResolver.ConfigChanged(lastBuild) ||
-		!equality.Semantic.DeepEqual(im.Env(), lastBuild.Spec.Env) ||
-		!equality.Semantic.DeepEqual(im.Resources(), lastBuild.Spec.Resources) ||
-		!equality.Semantic.DeepEqual(im.Bindings(), lastBuild.Spec.Bindings) {
-		reasons = append(reasons, v1alpha1.BuildReasonConfig)
+func (b *BuildDeterminer) IsBuildNeeded() (corev1.ConditionStatus, error) {
+	if !b.canDetermine() {
+		return corev1.ConditionUnknown, nil
 	}
 
-	if sourceResolver.RevisionChanged(lastBuild) {
-		reasons = append(reasons, v1alpha1.BuildReasonCommit)
+	var err error
+	b.changeSummary, err = buildchange.NewChangeProcessor().
+		Process(b.triggerChange()).
+		Process(b.commitChange()).
+		Process(b.configChange()).
+		Process(b.buildpackChange()).
+		Process(b.stackChange()).Summarize()
+	if err != nil {
+		return corev1.ConditionUnknown, err
 	}
 
-	if lastBuild.IsSuccess() {
-		if !builtWithBuildpacks(lastBuild, builder.BuildpackMetadata()) {
-			reasons = append(reasons, v1alpha1.BuildReasonBuildpack)
-		}
+	if b.changeSummary.HasChanges {
+		return corev1.ConditionTrue, nil
+	}
+	return corev1.ConditionFalse, nil
+}
 
-		if !builtWithStack(lastBuild, builder.RunImage()) {
-			reasons = append(reasons, v1alpha1.BuildReasonStack)
-		}
+func (b *BuildDeterminer) Reasons() string {
+	return b.changeSummary.ReasonsStr
+}
+
+func (b *BuildDeterminer) Changes() string {
+	return b.changeSummary.ChangesStr
+}
+
+func (b *BuildDeterminer) canDetermine() bool {
+	return b.srcResolver.Ready() && b.builder.Ready()
+}
+
+func (b *BuildDeterminer) lastBuildWasSuccessful() bool {
+	return b.lastBuild != nil && b.lastBuild.IsSuccess()
+}
+
+func (b *BuildDeterminer) triggerChange() buildchange.Change {
+	if b.lastBuild == nil || b.lastBuild.Annotations == nil {
+		return nil
 	}
 
-	if additionalBuildNeeded(lastBuild) {
-		reasons = append(reasons, v1alpha1.BuildReasonTrigger)
+	time, ok := b.lastBuild.Annotations[v1alpha1.BuildNeededAnnotation]
+	if !ok {
+		return nil
 	}
 
-	if len(reasons) == 0 {
-		return nil, corev1.ConditionFalse
+	return buildchange.NewTriggerChange(time)
+}
+
+func (b *BuildDeterminer) commitChange() buildchange.Change {
+	if b.lastBuild == nil || b.srcResolver.Status.Source.Git == nil {
+		return nil
 	}
 
-	return reasons, corev1.ConditionTrue
+	oldRevision := b.lastBuild.Spec.Source.Git.Revision
```

Comment:
What if the last build was not a git build?

Reply:
Aye. Will fix it.

Reply:
Couldn't the current build be a git build but the last build be a blob build for example?

Reply:
yikes... I definitely missed that. will fix. thanks for pointing it out

Reply:
resolved by https://github.com/pivotal/kpack/commit/be3ad6dd90eb514a1ce408e7c72bb7f649a74ff2